src/MembershipManagerInterface.php

```diff
@@ -55,6 +55,36 @@ public function getUserGroupIds(AccountInterface $user, array $states = [OgMembe
    */
   public function getUserGroups(AccountInterface $user, array $states = [OgMembershipInterface::STATE_ACTIVE]);
 
+  /**
+   * Returns an array of groups filtered by the og roles of the user.
+   *
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   The user to get the groups for.
+   * @param \Drupal\og\OgRoleInterface[] $roles
+   *   A list of og role objects to filter by.
+   * @param array $states
+   *   (optional) An array of states to filter the memberships by.
+   * @param bool $require_all_roles
+   *   (optional) If set to true, all requested roles must be present to return
+   *   the group.
+   */
+  public function getUserGroupsByRoles(AccountInterface $user, array $roles, array $states = [OgMembershipInterface::STATE_ACTIVE], bool $require_all_roles = FALSE);
```

Comment:
In the time since this PR was started we have decided to adopt PHP 7 type hints, so we can add the return type as well.